tdms/include/arrays/tensor3d.h

```diff
@@ -0,0 +1,149 @@
+/**
+ * @file tensor3d.h
+ * @author William Graham
+ * @brief template class for storing three-dimensional data arrays.
+ */
+#pragma once
+
+#include <vector>
+
+#include "cell_coordinate.h"
+
+/**
+ * @brief Template class for storing three-dimensional data as a strided vector.
+ * @details Three dimensional data is stored as a strided vector, in row-major
+ * (C) format. The last listed dimension has the fastest varying index, and the
+ * first listed dimension has the slowest listed. Explicitly,
+ *
+ * this->(i, j, k) = this[i*n_cols*n_layers + j*n_layers + k].
+ *
+ * This is consistent with the way hdf5 files store array-like data, see
+ * https://support.hdfgroup.org/HDF5/doc1.6/UG/12_Dataspaces.html.
+ * @tparam T Numerical datatype
+ */
+template<typename T>
+class Tensor3D {
+private:
+  /** @brief Convert a 3D (i,j,k) index to the corresponding index in the
+   * strided storage. */
+  int to_global_index(int i, int j, int k) const {
+    return i * n_layers_ * n_cols_ + j * n_layers_ + k;
+  }
+  int to_global_index(const ijk &index_3d) const {
+    return to_global_index(index_3d.i, index_3d.j, index_3d.k);
+  }
+
+protected:
+  int n_layers_ = 0;
+  int n_cols_ = 0;
+  int n_rows_ = 0;
+
+  /*! Strided vector that will store the array data */
+  std::vector<T> data_;
+
+  /*! The total number of elements, according to the dimension values currently
+   * set. */
+  int total_elements() const { return n_layers_ * n_cols_ * n_rows_; }
+
+public:
+  Tensor3D() = default;
+  Tensor3D(int n_layers, int n_cols, int n_rows) {
+    allocate(n_layers, n_cols, n_rows);
+  }
+  Tensor3D(T ***buffer, int n_layers, int n_cols, int n_rows) {
+    initialise(buffer, n_layers, n_cols, n_rows);
+  }
+
+  /** @brief Subscript operator for the Tensor, retrieving the (i,j,k)-th
+   * element. */
+  T &operator()(int i, int j, int k) { return data_[to_global_index(i, j, k)]; }
+  T operator()(int i, int j, int k) const {
+    return data_[to_global_index(i, j, k)];
+  }
+
+  /** @brief Subscript operator for the Tensor, retrieving the (i,j,k)-th
+   * element. */
+  T &operator[](const ijk &index_3d) {
+    return data_[to_global_index(index_3d)];
+  }
+  T operator[](const ijk &index_3d) const {
+    return data_[to_global_index(index_3d)];
+  }
+
+  /**
+   * @brief Whether the tensor contains any elements.
+   * @details Strictly speaking, checks whether the total_elements() method
+   * returns 0, indicating that this tensor has no size and thus no elements.
+   */
+  bool has_elements() const { return total_elements() != 0; }
```

Comment:
```suggestion
  bool is_nonzero() const { return total_elements() != 0; }
```

Maybe a better name? (Although you'd need to rename in VSCode).